### ISSUE TITLE
De-serialization bug from SavedModel format

### ISSUE THREAD

**Issue by fferroni** [tensorflow/tensorflow#42777]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.15.3
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Getting a ValueError when deserializing a model built with functional API and no custom layers.
```
ValueError: You tried to call `count_params` on input_mnist, but the layer isn't built. You can build it manually via: `input_mnist.build(batch_input_shape)`.
```

Test code to reproduce:

```
import os
import tempfile

import tensorflow as tf
import numpy as np


def test_bug_tensorflow():

    nb_inputs = 784
    hidden_units = [32, 32]
    nb_classes = 10

    # backbone model
    input_1 = tf.keras.layers.Input(shape=(nb_inputs,), name="input_backbone", dtype=tf.float32)
    x = input_1
    for i, units in enumerate(hidden_units):
        x = tf.keras.layers.Dense(units, name=f"dense_{i}", activation="relu")(x)
    backbone = tf.keras.models.Model(input_1, x, name="backbone")

    # classifier
    input_2 = tf.keras.layers.Input(shape=(nb_inputs,), name="input_mnist", dtype=tf.float32)
    x = backbone(input_2)
    x = tf.keras.layers.Dense(nb_classes, activation="softmax", name="output")(x)
    mnist_model = tf.keras.models.Model(input_2, x, name="mnist")
    mnist_model.predict(np.zeros((100, nb_inputs)))
    print(mnist_model.summary())

    tmp_dir = tempfile.TemporaryDirectory()
    tmp_path = str(tmp_dir.name)
    
    mnist_model.save(os.path.join(tmp_path, "model.h5"), save_format="h5")
    new_model = tf.keras.models.load_model(os.path.join(tmp_path, "model.h5"))
    mnist_model.predict(np.zeros((100, nb_inputs)))
    print(new_model.summary())

    mnist_model.save(tmp_path, save_format="tf")
    new_model = tf.keras.models.load_model(tmp_path)
    mnist_model.predict(np.zeros((100, nb_inputs)))
    print(new_model.summary())  # raises ValueError
```

**Describe the expected behavior**
Things work as expected with h5, but not save_format. If I remove the backbone model from the MNIST model, it works.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by fferroni** [comment]
Seems to be fixed in latest 2.3.0 release, however, I need to use v1.x 
Is there any way to patch this?

**Comment by Saduf2019** [comment]
@fferroni 
I ran the code on tf 1.15 and on tf-nightly and do not see any error in either, please refer tot he [gist here](https://colab.research.google.com/gist/Saduf2019/51f7fc8d0324b9d9003997b50959de93/untitled401.ipynb).
 Please let us know how can we help you.
